src/lib/lockfile.ts

```diff
@@ -0,0 +1,115 @@
+import * as fs from 'fs';
+import { dirname } from 'path';
+import { isRight } from 'fp-ts/lib/Either';
+
+import { exec, unlinkAll } from './fs-utils';
+import { NumericIdentifier } from '../types';
+
+// Equivalent to `drwxrwxrwt`
+const STICKY_WRITE_PERMISSIONS = 0o1777;
+
+const decodedUid = NumericIdentifier.decode(process.env.LOCKFILE_UID);
+export const LOCKFILE_UID = isRight(decodedUid) ? decodedUid.right : 65534;
+
+/**
+ * Internal lockfile manager to track files in memory
+ */
+// Track locksTaken, so that the proper locks can be cleaned up on process exit
+const locksTaken: { [lockName: string]: boolean } = {};
+
+// Returns all current locks taken, as they've been stored in-memory.
+// Optionally accepts filter function for only getting locks that match a condition.
+export const getLocksTaken = (
+	lockFilter: (path: string) => boolean = () => true,
+): string[] => Object.keys(locksTaken).filter(lockFilter);
+
+// Try to clean up any existing locks when the process exits
+process.on('exit', () => {
+	for (const lockName of getLocksTaken()) {
+		try {
+			unlockSync(lockName);
+		} catch (e) {
+			// Ignore unlocking errors
+		}
+	}
+});
+
+interface ChildProcessError {
+	code: number;
+	stderr: string;
+	stdout: string;
+}
+
+export class LockfileExistsError implements ChildProcessError {
+	public code: number;
+	public stderr: string;
+	public stdout: string;
+
+	constructor(path: string) {
+		this.code = 73;
+		this.stderr = `lockfile: Sorry, giving up on "${path}"`;
+		this.stdout = '';
+	}
+
+	// Check if an error is an instance of LockfileExistsError.
+	// This is necessary because the error thrown is a child process
+	// error that isn't typed by default, so instanceof will not work.
+	public static is(error: unknown): error is LockfileExistsError {
+		return (error as LockfileExistsError).code === 73;
+	}
+}
+
+export async function lock(path: string, uid = LOCKFILE_UID) {
+	/**
+	 * Set parent directory permissions to `drwxrwxrwt` (octal 1777), which are needed
+	 * for lockfile binary to run successfully as the `nobody` (UID 65534) user.
+	 * Otherwise the `nobody` user will not have write permissions to the necessary
+	 * bind-mounted lockfile directories. `chmod` does not fail or throw if the
+	 * directory already has the proper permissions.
+	 */
+	await fs.promises.chmod(dirname(path), STICKY_WRITE_PERMISSIONS);
```

Comment:
Maybe this needs to go into `update-lock.ts` as well, this does not make much sense in a generic module. 

Same thing with the `uid` argument, it should probably default to `os.userInfo().uid` and it is the responsibility of the caller to make sure that the given uid has write permissions on the directory.

Reply:
I made the default uid `os.userInfo().uid`, but I'm not sure about moving `STICKY_WRITE_PERMISSIONS` outside of lockfile.ts when it's only referenced in lockfile.ts. I settled on
```
if (uid !== 0) {
    await fs.promises.chmod(dirname(path), STICKY_WRITE_PERMISSIONS);
}
```
But otherwise moved BASE_LOCK_DIR and LOCKFILE_UID consts outside of lockfile. Thoughts?

Reply:
I think this is probably fine for now, since we are only using this in the supervisor and I don't want to keep blocking you

As a generic module, the following could potentially cause some issues
- `exec('<cmd>', uid)` will only succeed if `uid == os.userInfo().uid` or `os.userInfo().uid` is a privileged user. 
- On the first case, the call to `chmod` will fail unless the user already has the right permissions to the directory, in which case the call is useless (but benign probably)
- On the second case the call to `chmod` will succeed, but now the module has unilaterally changed the permissions of a directory, what if that directory is important? 

As I said, these considerations are thinking about this as an independent module, and if we ever move this somewhere else we will have to take that into account, but for now I think this works

Reply:
When thinking about lockfile as a independent module, the directory permissions change being unexpected or even undesired makes sense. I've added this as a consideration in the comments for lockfile, for future reference. Thanks!